src/wallet/wallet.cpp

```diff
@@ -1006,6 +1046,33 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockI
         if (fExisted && !fUpdate) return false;
         if (fExisted || IsMine(tx) || IsFromMe(tx))
         {
+            /* Check if any keys in the wallet keypool that were supposed to be unused
+             * have appeared in a new transaction. If so, remove those keys from the keypool.
+             * This can happen when restoring an old wallet backup that does not contain
+             * the mostly recently created transactions from newer versions of the wallet.
+             */
+            std::map<CKeyID, int64_t> keyPool = GetAllReserveKeys();
```

Comment:
Access the member directly to avoid a copy?

Also this will avoid running out of the keypool inside of one scriptPubKey :)